src/volatile-root/volatile-root.c

```diff
@@ -71,48 +74,74 @@ static int make_volatile(const char *root_path) {
 }
 
-static int make_overlay(const char *root_path) {
-        _cleanup_free_ char *escaped_path = NULL;
+static int make_overlay(const char *root_path, int overlay_type) {
+        _cleanup_free_ char *escaped_path = NULL, *usr_path = NULL;
         bool tmpfs_mounted = false;
-        const char *options = NULL;
+        const char *options = NULL, *overlay_name = NULL;
+        _cleanup_free_ char *mount_path = NULL;
+        _cleanup_free_ char *overlay_name_dir = NULL;
+        _cleanup_free_ char *upper_dir = NULL;
+        _cleanup_free_ char *work_dir = NULL;
         int r;
 
         assert(root_path);
 
-        r = mkdir_p("/run/systemd/overlay-sysroot", 0700);
+        if (overlay_type == VOLATILE_OVERLAY) {
+                overlay_name = overlay_root_name;
+                mount_path = strdup(root_path);
+                if (!mount_path)
+                        return log_oom();
+        } else {
+                assert(overlay_type == VOLATILE_OVERLAY_USR);
+                overlay_name = overlay_usr_name;
+                mount_path = strjoin(root_path, "/usr");
+        }
+
+        overlay_name_dir = strjoin("/run/systemd/", overlay_name);
+        upper_dir = strjoin(overlay_name_dir, "/upper");
+        work_dir = strjoin(overlay_name_dir, "/work");
+
+        r = mkdir_p(overlay_name_dir, 0700);
         if (r < 0)
-                return log_error_errno(r, "Couldn't create overlay sysroot directory: %m");
+                return log_error_errno(r, "Couldn't create overlay %s directory: %m", overlay_name);
 
-        r = mount_nofollow_verbose(LOG_ERR, "tmpfs", "/run/systemd/overlay-sysroot", "tmpfs", MS_STRICTATIME, "mode=0755" TMPFS_LIMITS_ROOTFS);
+        r = mount_nofollow_verbose(LOG_ERR, "tmpfs", overlay_name_dir, "tmpfs", MS_STRICTATIME, "mode=0755" TMPFS_LIMITS_ROOTFS);
         if (r < 0)
                 goto finish;
 
         tmpfs_mounted = true;
 
-        if (mkdir("/run/systemd/overlay-sysroot/upper", 0755) < 0) {
-                r = log_error_errno(errno, "Failed to create /run/systemd/overlay-sysroot/upper: %m");
+        if (mkdir(upper_dir, 0755) < 0) {
+                r = log_error_errno(errno, "Failed to create %s: %m", upper_dir);
                 goto finish;
         }
 
-        if (mkdir("/run/systemd/overlay-sysroot/work", 0755) < 0) {
-                r = log_error_errno(errno, "Failed to create /run/systemd/overlay-sysroot/work: %m");
+        if (mkdir(work_dir, 0755) < 0) {
+                r = log_error_errno(errno, "Failed to create %s: %m", work_dir);
                 goto finish;
         }
 
-        escaped_path = shell_escape(root_path, ",:");
+        if (overlay_type == VOLATILE_OVERLAY_USR) {
+                /* Create the usr directory in the sysroot if it doesn't exist */
+                r = chase("/usr", root_path, CHASE_PREFIX_ROOT, &usr_path, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "/usr not available in sysroot: %m");
```

Comment:
Hmm, the comment says the directory will be created, but the code doesn't match that?